### PR TITLE
generate header from yaml spec

### PR DESCRIPTION
**Pull request by rajveermalviya**
Add a custom yaml spec as discussed in #207, the details about the format can be found in the `schema.json`.

To test the generator:
1. First make sure you have Go installed.
1. Make some change in `webgpu.yml`
1. Run `make gen`

The generator can also be used to generate a separate implementation specific C header along side `webgpu.h`. The steps to do that is:
1. Create a new yaml file, for example `wgpu.yml`
1. Populate it with minimal required fields, refer to `schema.json` and [this example](https://gist.github.com/rajveermalviya/15200babba779bfc21d837af504a1860).
1. Run `go run ./gen -schema schema.json -yaml wgpu.yml -header wgpu.h`

Another option is to generate a single header from multiple yaml specs, which is also included but disabled currently.
I prefer separate headers because that's what `wgpu-native` uses currently.

### CODE REVIEW

File: gen/cheader.tmpl
Comment (ben-clayton) [inline gen/cheader.tmpl:101]: nit: parentheses are unnecessary here and below.

File: gen/cheader.tmpl
Comment (ben-clayton) [inline gen/cheader.tmpl:123]: Same thing about scoped indentation, here and below

File: gen/cheader.tmpl
Comment (ben-clayton) [inline gen/cheader.tmpl:131]:
```suggestion
{{- range $object := .Objects}}
```

here and below. If you accept this, you can fix all the occurrences with `{{- range $_,` -> `{{- range `

File: gen/cheader.tmpl
Comment (ben-clayton) [inline gen/cheader.tmpl:127]: What is the `{{"" -}}` for? You have this in a few places.

File: gen/gen.go
Comment (ben-clayton) [inline gen/gen.go:22]: An alternative to maintaining absolute indentations in the template is to have two functions to increment and decrement indentation. I've also used special characters in the template to increment and decrement indentation which is then processed at the end of the template.

Not asking for a change here, just suggesting alternatives for you to consider.

File: gen/gen.go
Comment (ben-clayton) [inline gen/gen.go:32]: FYI: It doesn't actually look like this is ever called.

File: gen/gen.go
Comment (ben-clayton) [inline gen/gen.go:35]: Template functions can return an error, so you don't have to use panic. e.g.

`func(s string, bitSize int) uint64` -> `func(s string, bitSize int) (uint64, error)`

File: gen/gen.go
Comment (ben-clayton) [inline gen/gen.go:50]:
```suggestion
			"IsLast": func(i int, s any) bool {
```

File: gen/gen.go
Comment (ben-clayton) [inline gen/gen.go:163]:
```suggestion
	sb := &strings.Builder{}
```

Then you can drop all the `&` on the `Fprintf` calls.

File: gen/utils.go
Comment (ben-clayton) [inline gen/utils.go:75]:
```suggestion
	for _, c := range s {
```

Which will correctly decode the runes from the string. You can then remove the `rune(c)` from below.

File: gen/utils.go
Comment (ben-clayton) [inline gen/utils.go:95]: ditto

File: gen/cheader.tmpl
Comment (rajveermalviya) [reply gen/cheader.tmpl]: I went through a pass of all `{{"\n" -}}` and removed all the ones that were unnecessary or replaceable.
Some are still there but for a reason, `{{"\n" -}}` were there to avoid unneeded newlines emitted from template. For example:

<table>
<tr>
<td>  </td>
<td> Template </td>
<td> Generates </td>
</tr>
<tr>
<td> 1. </td>
<td>

```go
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif{{"\n" -}}

{{if eq .Name "webgpu"}}
#include <stdint.h>
#include <stddef.h>
{{else}}
#include "webgpu.h"
{{end}}
```

</td>
<td>

```c
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif

#include <stdint.h>
#include <stddef.h>
```

</td>
</tr>

<tr>
<td> 2. </td>
<td>

```go
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif

{{if eq .Name "webgpu"}}
#include <stdint.h>
#include <stddef.h>
{{else}}
#include "webgpu.h"
{{end}}
```

</td>
<td>

```c
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif


#include <stdint.h>
#include <stddef.h>
```

</td>
</tr>

<tr>
<td> 3. </td>
<td>

```go
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif
{{if eq .Name "webgpu"}}
#include <stdint.h>
#include <stddef.h>
{{else}}
#include "webgpu.h"
{{end}}
```

</td>
<td>

```c
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif

#include <stdint.h>
#include <stddef.h>
```

</td>
</tr>

<tr>
<td> 4. </td>
<td>

```go
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif

{{- if eq .Name "webgpu"}}
#include <stdint.h>
#include <stddef.h>
{{else}}
#include "webgpu.h"
{{end}}
```

</td>
<td>

```c
#if !defined(WGPU_NULLABLE)
#define WGPU_NULLABLE
#endif
#include <stdint.h>
#include <stddef.h>
```

</td>
</tr>
</table>

Here only `1.` & `3.`  are producing desired the result, and `1.` allows us to have a newline in template, matching the generated result.

File: gen/cheader.tmpl
Comment (rajveermalviya) [reply gen/cheader.tmpl]: Similar reasons to `{{"\n" -}}` - to avoid unnecessary newlines. But now removed all instances.

File: gen/gen.go
Comment (rajveermalviya) [reply gen/gen.go]: `ParseUint` is used for enum & bitflags blocks.

Comment (ben-clayton) [reply gen/gen.go]: I see that it's used in the templates, but it doesn't ever appear to be called. You can replace the body with a `panic("!")` and see that it is never called.

**Review by ben-clayton — COMMENTED** [review]
Go / template readability LGTM. Thank you for all the changes.
Review still needed from project maintainers.

File: gen/gen.go
Comment (rajveermalviya) [reply gen/gen.go]: Correct, it'll only get called if in the yaml an enum/bitflag entry has a custom `value` property, which is optional.
Current `webgpu.yml` doesn't need them currently, but other custom specs may need them.

Comment (ben-clayton) [reply gen/gen.go]: 👍🏼 Just checking that this was intentional. SGTM

**Comment by kainino0x** [comment]
Sorry for the delay, I've been having a hard time catching up with work. Thank you Ben for the Go readability review!

One high level comment - I'd like to have a blocking GitHub Actions check that the generated output is up-to-date. This shouldn't be too complicated, just run the generator and then check that the git repository is clean.

**Comment by kainino0x** [comment]
(Oh, and - since we are generated files already - I wonder if we could also generate a JSON version of the YAML file, for easy ingestion by other projects? But let's save this for a future PR.)

File: gen/gen.go
Comment (kainino0x) [inline gen/gen.go:1]: I stumbled upon something - I remember you mentioned this would probably make GitHub say the project is written in Go.
Apparently we can do something like
`gen/** -linguist-detectable`
in .gitattributes to avoid this:
https://dev.to/katkelly/changing-your-repo-s-language-in-github-5gjo
https://github.com/github-linguist/linguist/blob/master/docs/overrides.md

File: schema.json
Comment (kainino0x) [inline schema.json:20]: (Can be future work)
Maybe we could use regex patterns to validate these a little more?
https://json-schema.org/understanding-json-schema/reference/string#regexp

These would get peppered all over the place so we'd probably want to link to the definition by reference:
https://stackoverflow.com/a/62585173/1268131

File: schema.json
Comment (kainino0x) [inline schema.json:24]: Regex patterns are useful here too: `^(0x[0-9A-F]{4}|0|[1-9][0-9]{0,4})$` or something like that

Though perhaps better - since we're using YAML already which supports hexadecimal numbers - could we just accept integers here instead? With range validation:
https://json-schema.org/understanding-json-schema/reference/numeric#range
(Can also be future work)

File: schema.json
Comment (kainino0x) [inline schema.json:262]: (Can be future work)
For the INIT macros (and generating C++ bindings), some optional members will need to know what their default value is (like WGPU_DEPTH_SLICE_UNDEFINED). Probably we can just put that here, though an alternative could be to move that information into the "type system" and give the `depthSlice` member a special type that knows its default.

File: schema.json
Comment (kainino0x) [inline schema.json:331]: (Can be future work)
This is used a bunch of times and I think could be defined by `$ref` as well (link again: https://stackoverflow.com/a/62585173/1268131)

`functions.items.properties.args` and `objects.items.properties.methods.items.properties.args` also could probably be the same, and maybe `structs.items.properties.members` too depending if it can be made to match (dawn does this too)

File: webgpu.h
Comment (kainino0x) [inline webgpu.h:428]: I think the YAML could specify a value for this so that it doesn't get renumbered? (We're going to need to renumber all these anyway later but it's convenient to minimize the differences for now)

File: gen/cheader.tmpl
Comment (kainino0x) [inline gen/cheader.tmpl:122]: nit: this blank line disappears from the generated result, but I think it would be more readable if it were kept

File: gen/cheader.tmpl
Comment (kainino0x) [inline gen/cheader.tmpl:127]: same for this blank line

**Review by kainino0x — APPROVED** [review]
I don't have much to say about the Go code. I don't know Go, but it seems good, and if it works, it works :) Looking at the schema, it looks really great!

The changes to webgpu.h look really easy to reproduce in Dawn (except the callback definition order will require a little work), so we can stay in sync until we start using this generator in Dawn (hopefully).

I added some comments for future work, but let's probably do those in another PR.
Just a few little things I'd like to do before landing this, but they're not critical.

File: schema.json
Comment (rajveermalviya) [reply schema.json]: default values are planned - In-order to keep this initial PR minimal, a future PR which includes both the header generation and default values populated in the yaml at once would be better.

File: gen/main.go
Comment (rajveermalviya) [reply gen/main.go]: removed for now, plan is to add it back later (for finding invalid duplicates across multiple yaml specs).

**Review by kainino0x — APPROVED** [review]
LGTM! Nice reduction on the schema, thanks for digging deeper into what we could do with ref.

**Comment by kainino0x** [comment]
I've left the future work comments "unresolved" as a reference.

**Comment by kainino0x** [comment]
I'll go ahead and merge since webgpu.h changes are minimal - in Dawn we can hopefully catch up soon so the sync process won't be made harder.

**Pull request merged**
